### PR TITLE
Add localization support

### PR DESCRIPTION
**Pull request by arkid15r**
Add python-holidays translation support.

Refactor existing code by @dragoon and @bkthomps (stacked on #810, #719):
  - Standardize country translation file lookup.
  - Change default language to None from "en" for compatibility.
  - Introduce `self.tr()` as a translation method.
  - Add `special_holidays` translation support.
  - Update `__eq__`,  `__ne__`, and `__str__` methods logic, fix tests.
  - Add holidays English translation for Argentina, Canada, Denmark, Poland, and Ukraine.
  - Add documentation on how to add new translations
  - Add .mo file generation mechanism
  - Update setup.cfg

### CODE REVIEW

File: holidays/holiday_base.py [hunk 1]
```diff
@@ -268,6 +282,22 @@ def __init__(
...
+                # Load translation.
+                try:
+                    translator = translation(
+                        name,
+                        languages=[language],
+                        localedir="locales",
+                    )
+                    self.tr = translator.gettext  # Replace `self.tr()`.
+                except FileNotFoundError:  # No translation found.
+                    pass
```
Comment (mborsetti) [inline holidays/holiday_base.py:299]: We should probably issue a warning at a minimum here, if not an outright Exception

**Review by mborsetti — COMMENTED** [review]
Great efforts!  A few suggestions and comments for the more senior people looking at this project.

File: holidays/holiday_base.py [hunk 2]
```diff
@@ -230,28 +241,31 @@ def __init__(
...
-
-        :param observed:
-            Whether to include the dates when public holiday are observed
-            (e.g. a holiday falling on a Sunday being observed the
-            following Monday). This doesn't work for all countries.
+        :param language:
+            The language which the returned holiday names will be translated
+            into. It must be an ISO 2-digit language code. If the language
+            translation is not supported the original holiday names will be
+            used.
```
[the remark this answers is not in the document]
Comment (dragoon) [reply holidays/holiday_base.py:248]: @mborsetti at the moment the library has a mix of english and local names for holidays, so probably that should be preserved for backwards compatibility.

File: holidays/countries/singapore.py [hunk 3]
```diff
@@ -58,6 +58,7 @@ def __init__(
         subdiv: Optional[str] = None,
         prov: Optional[str] = None,
         state: Optional[str] = None,
+        language: Optional[str] = None,
```
[the remark this answers is not in the document]
Comment (arkid15r) [reply holidays/countries/singapore.py:61]: This class (as well as MY) has its own __init__() signature. The new param is added to keep it consistent. Otherwise SG wouldn't have `language` available.

File: holidays/holiday_base.py [hunk 4]
```diff
@@ -230,28 +241,31 @@ def __init__(
...
-            Whether the entire year is calculated when one date from that year
-            is requested.
-
-        :param observed:
-            Whether to include the dates when public holiday are observed
-            (e.g. a holiday falling on a Sunday being observed the
-            following Monday). This doesn't work for all countries.
+        :param language:
+            The language which the returned holiday names will be translated
+            into. It must be an ISO 2-digit language code. If the language
```
[the remark this answers is not in the document]
Comment (arkid15r) [reply holidays/holiday_base.py:246]: Added, thank you!

File: holidays/holiday_base.py (hunk 1 above)
Comment (arkid15r) [reply holidays/holiday_base.py]: Good point, now it raises an exception.

File: holidays/holiday_base.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (arkid15r) [reply holidays/holiday_base.py]: First of all many thanks for the review, @mborsetti !

> I propose instead that we standardize the language returned when no language code is present.

I believe this approach is what Roman has originally implemented: use `en` as a default. I don't mind to re-work the logic back. My reasoning behind the `language = None` option is compatibility (learned the importance of this after many reviews by @dr-prodigy :) ). 
On the other hand it's kind of better UX to have holiday names in the user's language by default (instead of English).

>I further propose that such standard be that this project behaves the same way as gettext, i.e. the language returned is based on the "current global domain, language, and locale directory". However, if no match is found, then en language will be returned.

That's an interesting idea. I'll look into it.

I admit that sometimes (quite often) it's impossible to go with the right approach due to compatibility reasons.

**Comment by dr-prodigy** [comment]
Hi @arkid15r @dragoon @mborsetti , thank you all for your valuable work and considerations!

I agree on the fact that "original language" makes no sense, but:
up to now, we have been supporting only one language per country.. often English, sometimes another (typically the official / most important language of that country).

This given, I think we should not aim neither to default to the system / user locale (it would be only seldom available) nor always to English (this would require a massive translation work for all non English-speaking countries).

Instead, the easiest approach (not far from that "original language" concept) would then be to:
- add an explicit default language setting per each country (similar to what we do with subdivisions.. it also should be documented in the country list)
- set that default to the currently existing language
- for multi-language countries, add any other required language

This done, whenever a holiday object is instantiated, the default would then be used both if:
- no language is specified
- if the specified language is not available.
Basically, minimum effort + no behaviour changes.

In a second phase, English could even possibly be added for any country, if desired.
This long-running task, although beneficial for standardization, would anyway be completely optional and procrastinable.

IMHO this would allow for the quickest implementation and the less possible impact in terms of compatibility.

What do you all think?

Thx again everyone, KUTGW! 👍

File: holidays/holiday_base.py (hunk 1 above)
Comment (arkid15r) [reply holidays/holiday_base.py]: Replaced with a warning after another thought.

**Comment by mborsetti** [comment]
@dr-prodigy Sounds like a great plan!

I propose that the documentation is augmented with both the default language (ISO 639-1 2-letter language code) and the alternative languages supported, plus an explanation of the behavior (pretty much what you said above).

File: holidays/countries/singapore.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mborsetti) [reply holidays/countries/singapore.py:61]: Oh, yeah! Now I remember, I did it early on in the project in order to surface the docstring in IDEs. This was before we standardized on using `holidays.country_holidays()`, which provides docstring no matter what country you call.

I will edit it away in a separate, forthcoming, PR.

Thanks and apologies.

File: holidays/holiday_base.py [hunk 5]
```diff
@@ -545,10 +576,24 @@ def pop_named(self, name: str) -> List[date]:
         return to_pop
 
     def __eq__(self, other: object) -> bool:
-        return dict.__eq__(self, other) and self.__dict__ == other.__dict__
+        that = copy.deepcopy(other)
```
[the remark this answers is not in the document]
Comment (mborsetti) [reply holidays/holiday_base.py]: Now I am even more confused: per the documentation, `copy.copy()` and `dict.copy()` should both create _shallow_ copies, so why is one working and the other not?  

And this https://stackoverflow.com/a/72164590/12228585 says that they are "they are functionally equivalent" and point to an inspection of the copy.copy() code.

Are you sure that `dict.copy()` doesn't work?

**Comment by dr-prodigy** [comment]
@mborsetti Yes Mike, totally agree 👍

File: holidays/holiday_base.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (arkid15r) [reply holidays/holiday_base.py]: I believe the reason is that we have dict-based objects (not dicts). Yes, dict.copy() creates a shallow copy of a dict. But we need other attributes too.

I'll look into this later. There should be a way to improve it a bit.

**Comment by arkid15r** [comment]
Mauri and Mike, thanks for the ideas!

I think that the original language makes perfect sense to users as long as it's a language they understand natively; implementation  examples: [PL](https://github.com/dr-prodigy/python-holidays/blob/master/holidays/countries/poland.py) or [JP](https://github.com/dr-prodigy/python-holidays/blob/master/holidays/countries/japan.py).

I agree that the default language as well as fallback to user locale based language may lead to an unexpected behaviour, compatibility issues, and may require a lot of translation efforts to make it right.

My understanding is that@dr-prodigy's approach differs from what we have at the moment just by an extra `default_language` parameter (which is not required for the implementation as the original holiday names language is an obvious default). However, it definitely should be added to the country documentation table together with the list of available translation.

My questions: 
  * what to do with cases like [AR](https://github.com/dr-prodigy/python-holidays/blob/master/holidays/countries/argentina.py) where holiday names already include English translation. We could remove English from the names and add them to a separate translation. There is a slight chance of introducing a breaking change for users who rely on the current format. But after fixing that we'd have a standardized files everywhere.
 
  * on what stage .mo files will be generated? The binary files are difficult to review so auto-generation upon build/test looks appealing. The user generated files approach is easier to implement though.

**Comment by dr-prodigy** [comment]
> Mauri and Mike, thanks for the ideas!
> 
> I think that the original language makes perfect sense to users as long as it's a language they understand natively; implementation examples: [PL](https://github.com/dr-prodigy/python-holidays/blob/master/holidays/countries/poland.py) or [JP](https://github.com/dr-prodigy/python-holidays/blob/master/holidays/countries/japan.py).
> 
> I agree that the default language as well as fallback to user locale based language may lead to an unexpected behaviour, compatibility issues, and may require a lot of translation efforts to make it right.
> 
> My understanding is that @dr-prodigy's approach differs from what we have at the moment just by an extra `default_language` parameter (which is not required for the implementation as the original holiday names language is an obvious default). However, it definitely should be added to the country documentation table together with the list of available translation.
Yes you're right: the parameter is not strictly required per se (as a default, you can just default to it :-) ), but it could become useful if we want - for instance - to add a warning in case user asks for a not existing language (if he asks explicitly for the default language, no warning would apply). 
> 
> My questions:
> 
> * what to do with cases like [AR](https://github.com/dr-prodigy/python-holidays/blob/master/holidays/countries/argentina.py) where holiday names already include English translation. We could remove English from the names and add them to a separate translation. There is a slight chance of introducing a breaking change for users who rely on the current format. But after fixing that we'd have a standardized files everywhere.
I agree. Being descriptions, I don't feel this like a breaking change. I'm definitely towards its standardization (ie: remove double language, add English).
> * on what stage .mo files will be generated? The binary files are difficult to review so auto-generation upon build/test looks appealing. The user generated files approach is easier to implement though.
Definitely agree on this too: I'd like to see their build automated.. what about pre-commit phase?

**Comment by mborsetti** [comment]
> I agree that the default language as well as fallback to user locale based language may lead to an unexpected behaviour, compatibility issues, and may require a lot of translation efforts to make it right.

I thought I'd clarify my initial standardization proposal. 

Using the language "based on the current global domain, language, and locale directory" is the default, out-of-the-box, behavior of `gettext`, the Python module we use. Indeed, I copied this description word by word from [the official Python documentation](https://docs.python.org/3/library/gettext.html#:~:text=return%20the%20localized%20translation%20of%20message%2C%20based%20on%20the%20current%20global%20domain%2C%20language%2C%20and%20locale%20directory), so there's no unexpected behavior. An unexpected behavior for me would be a `gettext`-based project that behaves in a different, non-standard, way.

In my own experience, this is also how localization is implemented IRL:

- Install Chrome or whatever on a computer with the OS in Italian, and the menus, error messages, etc. are in Italian without having to do any gymnastics to get there. 
- Install Chrome or whatever on a computer with the OS in Italian but want to have the app in German? Then you must explicitly specify the language (by doing gymnastics, e.g. fiddling with settings to switch it or passing a CLI argument at startup). 
- Install Chrome or whatever on a machine with an _unsupported_ language, then everything is in its default language, i.e. English (or the installer show a language picker). 

@dr-prodigy as usual I defer to you and thanks for all the work!

**Comment by dragoon** [comment]
Thanks @arkid15r, the PR looks good to me.

I have included the .mo files in my original PR since that was the easiest option indeed. Ithink, however, that they should be generated as a part of the build process. The `python-i18n` library from the PR #719 was easier in this regard :)
I can try to look into how to add this to your current github actions setup if you agree on the approach.

**Comment by arkid15r** [comment]
> Thanks @arkid15r, the PR looks good to me.
> 
> I have included the .mo files in my original PR since that was the easiest option indeed. Ithink, however, that they should be generated as a part of the build process. The `python-i18n` library from the PR #719 was easier in this regard :) I can try to look into how to add this to your current github actions setup if you agree on the approach.

Thank you for reviewing this, Roman!

I agree it could have been somewhat easier to implement it using the `python-i18n` but we've decided to avoid an extra dependency and went with the `gettext` approach. I'm going to look into the .mo auto-generation this weekend. There may be another case - when those files are required for testing purposes.

I highly appreciate your willingness help!

**Comment by arkid15r** [comment]
Just to double-check I'm not missing anything. 

The items I'm going to address in following updates:
  * `default_language` attribute (per Mauri's suggestion)
  * mixed language holiday names split example (AR)
  * .mo files automatic generation (it seems we need those for tests and package build phases only)
  * `gettext`-like behavior (as Mike suggested)
  * add translation instructions, document .pot/.po/.mo generation steps

**Comment by arkid15r** [comment]
> The items I'm going to address in following updates:
> 
> * `default_language` attribute (per Mauri's suggestion)
> * mixed language holiday names split example (AR)
> * .mo files automatic generation (it seems we need those for tests and package build phases only)
> * `gettext`-like behavior (as Mike suggested)
> * add translation instructions, document .pot/.po/.mo generation steps

Finally looks green! I pushed the i18n related updates and fixed the tests in several attempts.

Please review and test locally if possible.
